Spring Data R2DBC sample broke with upgrade to Spring Boot 2.5.1

In `SpringDataR2dbcApp`, the Spring Data sample programmatically instantiates a `ConnectionFactory` object. This used to work, with Spring Boot creating `DatabaseClient` and repository objects based on the application-provided `ConnectionFactory`.

However, after upgrade to Spring Boot 2.5.1, the sample failed to start at autoconfiguration with `Caused by: java.lang.IllegalArgumentException: Cannot determine database's type as ConnectionFactory is not options-capable`. This issue goes away in Spring Boot 2.5.2; however, I'd expect it to keep breaking, as creating a `ConnectionFactory` manually is no longer the recommended Spring Boot approach.

We have two options:
1) Configure the data source with a property instead of programmatically. I prefer this, since that's how the majority of our users will be using R2DBC with Spring Boot.
2) Use Spring Boot's  `ConnectionFactoryBuilder`, as described in [this comment](https://github.com/spring-projects/spring-boot/issues/26977#issuecomment-864369923)

--- a/cloud-spanner-r2dbc-samples/cloud-spanner-spring-data-r2dbc-sample/pom.xml
+++ b/cloud-spanner-r2dbc-samples/cloud-spanner-spring-data-r2dbc-sample/pom.xml
@@ -42,10 +42,52 @@
       <artifactId>spring-boot-starter-webflux</artifactId>
     </dependency>
 
+    <!-- Test dependencies -->
+    <dependency>
+      <groupId>org.springframework.boot</groupId>
+      <artifactId>spring-boot-starter-test</artifactId>
+      <scope>test</scope>
+    </dependency>
   </dependencies>
 
   <build>
     <plugins>
+
+      <!-- This sample inherits from Spring Boot Starter, and not from the project pom,
+           so configuration separating unit from integration tests needs to be duplicated.
+      -->
+      <plugin>
+        <groupId>org.apache.maven.plugins</groupId>
+        <artifactId>maven-surefire-plugin</artifactId>
+        <configuration>
+          <skip>${skip.surefire.tests}</skip>
+          <excludes>
+            <exclude>**/*IntegrationTest*</exclude>
+          </excludes>
+        </configuration>
+      </plugin>
+
+      <plugin>
+        <groupId>org.apache.maven.plugins</groupId>
+        <artifactId>maven-failsafe-plugin</artifactId>
+
+        <configuration>
+          <includes>
+            <include>**/*IntegrationTest*</include>
+          </includes>
+        </configuration>
+
+        <executions>
+          <execution>
+            <id>integration-tests</id>
+            <goals>
+              <goal>integration-test</goal>
+              <goal>verify</goal>
+            </goals>
+          </execution>
+        </executions>
+
+      </plugin>
 
       <plugin>
         <artifactId>maven-deploy-plugin</artifactId>

--- a/cloud-spanner-r2dbc-samples/cloud-spanner-spring-data-r2dbc-sample/src/main/java/com/example/SpringDataR2dbcApp.java
+++ b/cloud-spanner-r2dbc-samples/cloud-spanner-spring-data-r2dbc-sample/src/main/java/com/example/SpringDataR2dbcApp.java
@@ -16,17 +16,11 @@
 
 package com.example;
 
-import static com.google.cloud.spanner.r2dbc.SpannerConnectionFactoryProvider.DRIVER_NAME;
 import static com.google.cloud.spanner.r2dbc.SpannerConnectionFactoryProvider.INSTANCE;
 import static io.r2dbc.spi.ConnectionFactoryOptions.DATABASE;
-import static io.r2dbc.spi.ConnectionFactoryOptions.DRIVER;
 import static org.springframework.web.reactive.function.server.RequestPredicates.GET;
 import static org.springframework.web.reactive.function.server.RouterFunctions.route;
 
-import io.r2dbc.spi.ConnectionFactories;
-import io.r2dbc.spi.ConnectionFactory;
-import io.r2dbc.spi.ConnectionFactoryOptions;
-import io.r2dbc.spi.Option;
 import java.net.URI;
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
@@ -69,20 +63,6 @@ public class SpringDataR2dbcApp {
     Assert.notNull(GCP_PROJECT, "Please provide gcp.project property");
 
     SpringApplication.run(SpringDataR2dbcApp.class, args);
-  }
-
-
-  @Bean
-  public static ConnectionFactory spannerConnectionFactory() {
-    ConnectionFactory connectionFactory = ConnectionFactories.get(ConnectionFactoryOptions.builder()
-        .option(Option.valueOf("project"), GCP_PROJECT)
-        .option(DRIVER, DRIVER_NAME)
-        .option(INSTANCE, SPANNER_INSTANCE)
-        .option(DATABASE, SPANNER_DATABASE)
-        .option(Option.valueOf("client-implementation"), "client-library")
-        .build());
-
-    return connectionFactory;
   }
 
   @EventListener(ApplicationReadyEvent.class)

--- a/cloud-spanner-r2dbc-samples/cloud-spanner-spring-data-r2dbc-sample/src/main/resources/application.properties
+++ b/cloud-spanner-r2dbc-samples/cloud-spanner-spring-data-r2dbc-sample/src/main/resources/application.properties
@@ -1,1 +1,4 @@
+spring.r2dbc.url=\
+r2dbc:cloudspanner://spanner.googleapis.com:443/projects/${gcp.project}/instances/${spanner.instance}/databases/${spanner.database}
+
 logging.level.org.springframework.r2dbc=DEBUG

--- a/cloud-spanner-r2dbc-samples/cloud-spanner-spring-data-r2dbc-sample/src/test/java/com/example/SpringDataR2dbcAppIntegrationTest.java
+++ b/cloud-spanner-r2dbc-samples/cloud-spanner-spring-data-r2dbc-sample/src/test/java/com/example/SpringDataR2dbcAppIntegrationTest.java
@@ -1,0 +1,55 @@
+package com.example;
+
+import static org.assertj.core.api.Assertions.assertThat;
+
+import com.google.cloud.ServiceOptions;
+import java.util.concurrent.atomic.AtomicReference;
+import org.junit.jupiter.api.Test;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.boot.test.context.SpringBootTest;
+import org.springframework.boot.test.context.SpringBootTest.WebEnvironment;
+import org.springframework.test.context.DynamicPropertyRegistry;
+import org.springframework.test.context.DynamicPropertySource;
+import org.springframework.test.web.reactive.server.WebTestClient;
+import reactor.core.publisher.Mono;
+
+@SpringBootTest(webEnvironment = WebEnvironment.RANDOM_PORT)
+public class SpringDataR2dbcAppIntegrationTest {
+
+  @DynamicPropertySource
+  static void registerProperties(DynamicPropertyRegistry registry) {
+    registry.add("gcp.project", () -> ServiceOptions.getDefaultProjectId());
+    registry.add("spanner.database", () -> System.getProperty("spanner.database","testdb"));
+    registry.add("spanner.instance", () -> System.getProperty("spanner.instance", "reactivetest"));
+  }
+
+  @Autowired
+  private WebTestClient webTestClient;
+
+  @Test
+  public void testAllWebEndpoints() {
+    
+    // initially empty table
+    this.webTestClient.get().uri("/list").exchange()
+        .expectBody(Book[].class).isEqualTo(new Book[0]);
+
+    this.webTestClient.post().uri("/add").body(Mono.just("Call of the wild"), String.class)
+        .exchange().expectStatus().is2xxSuccessful();
+
+    AtomicReference<String> id = new AtomicReference<>();
+    this.webTestClient.get().uri("/list").exchange()
+        .expectBody(Book[].class).value(books -> {
+       assertThat(books).hasSize(1);
+       assertThat(books[0].getTitle()).isEqualTo("Call of the wild");
+       id.set(books[0].getId());
+    });
+
+    assertThat(id).doesNotHaveValue("");
+
+    this.webTestClient.get().uri("/search/"  +id.get()).exchange()
+        .expectBody(Book.class).value(book -> {
+      assertThat(book.getTitle()).isEqualTo("Call of the wild");
+    });
+  }
+
+}

